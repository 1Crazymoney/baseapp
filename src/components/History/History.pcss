.pg-history {
  background: var(--base-bright);

  td {
    &:first-child {
      min-width: 100px;
    }
  }
}

.pg-history--loading {
  display: flex;
  justify-content: center;
  align-items: center;
  padding: calc(var(--gap) * 3);
}

.pg-history--empty {
  .cr-table__body {
    tr {
      td:last-child {
        display: block;
        width: 100%;
        font-size: 11px !important;
        font-weight: bold;
        margin-left: 10px;
        padding: 20px 0 20px 0 !important;
      }
    }
  }
}

.pg-history--loading {
  display: flex;
  justify-content: center;
  align-items: center;
  padding: calc(var(--gap) * 3);
}

.pg-history--empty {
  .cr-table__body {
    tr {
      td:last-child {
        display: block;
        width: 100%;
        font-size: 11px !important;
        font-weight: bold;
        margin-left: 10px;
        padding: 20px 0 20px 0 !important;
      }
    }
  }
}
