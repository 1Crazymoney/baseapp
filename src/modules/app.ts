import { combineReducers } from 'redux';
import { authReducer, AuthState } from './auth';
import { contactReducer, ContactState } from './contact';
import { currencyHistoryReducer, CurrencyHistoryState } from './history/currencyHistory';
import { depositsReducer, DepositsState } from './history/deposits';
import { PrivateTradesState, tradesReducer } from './history/trades';
import { withdrawsReducer, WithdrawsState } from './history/withdraws';
import {
    documentsReducer,
    DocumentsState,
    identityReducer,
    IdentityState,
    phoneReducer,
    PhoneState,
} from './kyc';
import { marketsReducer, MarketsState } from './markets';
import {
    depthReducer,
    DepthState,
    orderBookReducer,
    OrderBookState,
} from './orderBook';
import { ordersReducer, OrdersState } from './orders';
import { passwordReducer, PasswordState } from './password';
import { profileReducer, ProfileState } from './profile';
import { rangerReducer, RangerState } from './ranger/reducer';
import { recentTradesReducer, RecentTradesState } from './recentTrades';
import { userActivityReducer, UserActivityState } from './userActivity';
import { walletsReducer, WalletsState } from './wallets';

export interface AppState {
    auth: AuthState;
    contact: ContactState;
    deposits: DepositsState;
    trades: PrivateTradesState;
    recentTrades: RecentTradesState;
    withdraws: WithdrawsState;
    orders: OrdersState;
    password: PasswordState;
    profile: ProfileState;
    wallets: WalletsState;
    documents: DocumentsState;
    identity: IdentityState;
    phone: PhoneState;
    markets: MarketsState;
    orderBook: OrderBookState;
    depth: DepthState;
    currencyHistory: CurrencyHistoryState;
    userActivity: UserActivityState;
    ranger: RangerState;
}

export const appReducer = combineReducers({
    auth: authReducer,
    contact: contactReducer,
    deposits: depositsReducer,
    trades: tradesReducer,
    recentTrades: recentTradesReducer,
    withdraws: withdrawsReducer,
    orders: ordersReducer,
    password: passwordReducer,
    profile: profileReducer,
    wallets: walletsReducer,
    phone: phoneReducer,
    identity: identityReducer,
    documents: documentsReducer,
    markets: marketsReducer,
    orderBook: orderBookReducer,
    depth: depthReducer,
    currencyHistory: currencyHistoryReducer,
    userActivity: userActivityReducer,
    ranger: rangerReducer,
});
