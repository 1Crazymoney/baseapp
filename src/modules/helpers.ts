import { MakerType } from './history';

const makerTypeMap = {
    ask: 'sell',
    bid: 'buy',
};

export const kindToMakerType = (kind: string): MakerType => makerTypeMap[kind];
