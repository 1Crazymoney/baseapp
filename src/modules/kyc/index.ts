import { DocumentsState } from './documents';
import { IdentityState } from './identity';
import { PhoneState } from './phone';

export * from './documents';
export * from './identity';
export * from './phone';


export type KycState = PhoneState & IdentityState & DocumentsState;
