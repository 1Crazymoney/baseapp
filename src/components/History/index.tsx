import { History, Loader } from '@openware/components';
import classnames from 'classnames';
import * as React from 'react';
import { connect, MapDispatchToPropsFunction } from 'react-redux';
import { localeDate } from '../../helpers';
import { RootState, selectTradesLoading, selectTradesOfCurrentMarket, Trade } from '../../modules';
import { tradesFetch } from '../../modules/history/trades/actions';
import {
    Market,
    selectCurrentMarket,
} from '../../modules/markets';

interface ReduxProps {
    trades: Trade[];
    currentMarket: Market;
    ordersLoading?: boolean;
}

interface DispatchProps {
    getTradesHistory: typeof tradesFetch;
}

interface HistoryState {
    historyData: Trade[];
}

type Props = ReduxProps & DispatchProps;
const sides = ['buy', 'sell'];

class HistoryContainer extends React.Component<Props, HistoryState> {
    public componentDidMount() {
        this.props.getTradesHistory([this.props.currentMarket]);
    }

    public componentWillReceiveProps(next: Props) {
        if (this.props.currentMarket !== next.currentMarket) {
            this.props.getTradesHistory([next.currentMarket]);
        }
    }

    public render() {
        const { ordersLoading, trades } = this.props;
        const classNames = classnames('pg-history', {
            'pg-history--empty': !trades.length,
            'pg-history--loading': ordersLoading,
        });
        return (
            <div className={classNames}>
                {ordersLoading ? <Loader /> : this.history()}
            </div>
        );
    }

    private history = () => <History headers={this.renderHeaders()} data={this.renderData()} />;

    private static getDate(time) {
        return localeDate(time);
    }

    private convertTotal(total: number, fractionDigit = 6) {
        return +Number(total).toFixed(fractionDigit);
    }

    private renderHeaders() {
      return ['Date', 'Action', 'Price', 'Amount', 'Total'];
    }

    private renderSide(side) {
        if (!!(sides.indexOf(side) + 1)) {
            return side === 'sell' ? 'ask' : 'bid';
        }

        return side;
    }

    private renderData() {
        const { trades } = this.props;
        return (trades.length > 0) ? trades.map(item => {
            const {
                created_at,
                side,
                price,
                volume,
            } = item;

            const resultSide = this.renderSide(side);
            return [
                HistoryContainer.getDate(created_at),
                resultSide,
                price,
                volume,
                this.convertTotal(Number(volume) * Number(price)),
            ];
        }) : [['There is no data to show...']];
    }
}

const mapStateToProps = (state: RootState): ReduxProps => ({
    trades: selectTradesOfCurrentMarket(state),
    ordersLoading: selectTradesLoading(state),
    currentMarket: selectCurrentMarket(state),
});

const mapDispatchToProps: MapDispatchToPropsFunction<DispatchProps, {}> =
    dispatch => ({
        getTradesHistory: (markets: Market[]) => dispatch(tradesFetch(markets)),
    });

export const HistoryComponent = connect(mapStateToProps, mapDispatchToProps)(HistoryContainer);
