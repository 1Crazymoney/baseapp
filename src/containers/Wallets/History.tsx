import {
    Decimal,
    History,
    Pagination,
    WalletItemProps,
} from '@openware/components';
import * as moment from 'moment';
import * as React from 'react';
import {
    InjectedIntlProps,
    injectIntl,
    intlShape,
} from 'react-intl';
import { connect, MapDispatchToPropsFunction } from 'react-redux';
import { localeDate } from '../../helpers';
import {
    fetchHistory,
    List,
    resetHistory,
    RootState,
    selectCurrentPage,
    selectFirstElemIndex,
    selectFullHistory,
    selectHistory,
    selectHistoryLoading,
    selectLastElemIndex,
    selectNextPageExists,
    selectPageCount,
    selectWallets,
} from '../../modules';
import { FailIcon } from './FailIcon';
import { SucceedIcon } from './SucceedIcon';

export interface HistoryProps {
    label: string;
    type: string;
    currency: string;
}

export interface ReduxProps {
    list: List;
    wallets: WalletItemProps[];
    fetching: boolean;
    fullHistory: number;
    page: number;
    pageCount: number;
    firstElemIndex: number;
    lastElemIndex: number;
    nextPageExists: boolean;
}

interface DispatchProps {
    fetchHistory: typeof fetchHistory;
    resetHistory: typeof resetHistory;
}

export type Props = HistoryProps & ReduxProps & DispatchProps & InjectedIntlProps;

export class WalletTable extends React.Component<Props> {
    //tslint:disable-next-line:no-any
    public static propTypes: React.ValidationMap<any> = {
        intl: intlShape.isRequired,
    };

    public componentDidMount() {
        const { type, currency } = this.props;
        this.props.fetchHistory({ page: 0, currency, type, limit: 6 });
    }

    public componentWillReceiveProps(nextProps) {
        const { type, currency } = this.props;
        if (nextProps.currency !== currency || nextProps.type !== type) {
            this.props.resetHistory();
            this.props.fetchHistory({ page: 0, currency: nextProps.currency, type, limit: 6 });
        }
    }

    public componentWillUnmount() {
        this.props.resetHistory();
    }
    // tslint:disable
    public render() {
        const { label, list, fullHistory, firstElemIndex, lastElemIndex, page, nextPageExists } = this.props;

        if (!list.length) {
            return null;
        }
        return (
            <div className="pg-history-elem__wallet">
                <div className="pg-history-elem__label">
                    {this.props.intl.formatMessage({ id: `page.body.history.${label}` })}
                </div>
                <History headers={this.getHeaders(label)} data={this.retrieveData(list)}/>
                <Pagination
                    firstElemIndex={firstElemIndex}
                    lastElemIndex={lastElemIndex}
                    total={fullHistory}
                    page={page}
                    nextPageExists={nextPageExists}
                    onClickPrevPage={this.onClickPrevPage}
                    onClickNextPage={this.onClickNextPage}
                />
            </div>
        );
    }

    private getHeaders = (label: string) => [
        this.props.intl.formatMessage({ id: `page.body.history.${label}.header.date`}),
        this.props.intl.formatMessage({ id: `page.body.history.${label}.header.status`}),
        this.props.intl.formatMessage({ id: `page.body.history.${label}.header.amount`}),
    ];

    private onClickPrevPage = () => {
        const { page, type, currency } = this.props;
        this.props.fetchHistory({ page: page - 1, currency, type, limit: 6 });
    };

    private onClickNextPage = () => {
        const { page, type, currency } = this.props;
        this.props.fetchHistory({ page: page + 1, currency, type, limit: 6 });
    };

    private retrieveData = list => {
        const currentWallet = this.props.wallets.find(w => w.currency === this.props.currency)
            || { fixed: 8 };
        if (list.length === 0) {
            return [[this.props.intl.formatMessage({ id: 'page.noDataToShow'}), '', '']];
        }
        return [...list]
            .sort((a, b) => {
                return moment(localeDate(a.created_at), 'DD/MM HH:mm') > moment(localeDate(b.created_at), 'DD/MM HH:mm') ? -1 : 1;
            })
            .map(item => {
                return [
                    moment(item.created_at).format('DD-MM YYYY'),
                    this.formatTxState(item.state),
                    <Decimal fixed={currentWallet.fixed}>{item.amount}</Decimal>,
                ];
            });
    };

    private formatTxState = (tx: string) => {
        const statusMapping = {
            succeed: <SucceedIcon/>,
            failed: <FailIcon/>,
            accepted: <SucceedIcon/>,
            collected: <SucceedIcon/>,
            canceled: <FailIcon/>,
            rejected: <FailIcon/>,
            processing: this.props.intl.formatMessage({ id: 'page.body.wallets.table.pending'}),
            prepared: this.props.intl.formatMessage({ id: 'page.body.wallets.table.pending'}),
            submitted: this.props.intl.formatMessage({ id: 'page.body.wallets.table.pending'}),
            skipped:  <SucceedIcon/>,
        };
        return statusMapping[tx];
    };
}


export const mapStateToProps = (state: RootState): ReduxProps => ({
    list: selectHistory(state),
    wallets: selectWallets(state),
    fetching: selectHistoryLoading(state),
    fullHistory: selectFullHistory(state),
    page: selectCurrentPage(state),
    pageCount: selectPageCount(state, 6),
    firstElemIndex: selectFirstElemIndex(state, 6),
    lastElemIndex: selectLastElemIndex(state, 6),
    nextPageExists: selectNextPageExists(state, 6),
});

export const mapDispatchToProps: MapDispatchToPropsFunction<DispatchProps, {}> =
    dispatch => ({
        fetchHistory: params => dispatch(fetchHistory(params)),
        resetHistory: () => dispatch(resetHistory()),
    });

export const WalletHistory = injectIntl(connect(mapStateToProps, mapDispatchToProps)(WalletTable));
