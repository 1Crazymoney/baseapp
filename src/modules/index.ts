import { combineReducers } from 'redux';
// tslint:disable-next-line no-submodule-imports
import { all, call } from 'redux-saga/effects';
import { appReducer, AppState } from './app';
import { rootAuthSaga } from './auth';
import { rootSendEmailSaga } from './contact';
import { rootHandleErrorSaga } from './error';
import { rootCurrencyHistorySaga } from './history/currencyHistory';
import { rootDepositSaga } from './history/deposits';
import { rootTradeSaga } from './history/trades';
import { rootWithdrawSaga } from './history/withdraws';
import { rootSendDocumentsSaga } from './kyc/documents';
import { rootSendIdentitySaga } from './kyc/identity';
import { rootSendCodeSaga } from './kyc/phone';
import { rootMarketsSaga } from './markets';
import { rootOrderBookSaga } from './orderBook';
import { rootOrdersSaga } from './orders';
import { rootPasswordSaga } from './password';
import { rootProfileSaga } from './profile';
import { rootRecentTradesSaga } from './recentTrades';
import { rootUserActivitySaga } from './userActivity';
import { rootWalletsSaga } from './wallets';

export * from './auth';
export * from './contact';
export * from './wallets';
export * from './profile';
export * from './markets';
export * from './orderBook';
export * from './orders';
export * from './password';
export * from './userActivity';

export * from './history/currencyHistory';
export * from './history/deposits';
export * from './history/trades';
export * from './history/withdraws';
export * from './kyc';

export * from './error';

export interface RootState {
    app: AppState;
}

export const rootReducer = combineReducers({
    app: appReducer,
});

export function* rootSaga() {
    yield all([
        call(rootAuthSaga),
        call(rootDepositSaga),
        call(rootTradeSaga),
        call(rootWithdrawSaga),
        call(rootMarketsSaga),
        call(rootOrdersSaga),
        call(rootProfileSaga),
        call(rootWalletsSaga),
        call(rootPasswordSaga),
        call(rootSendCodeSaga),
        call(rootSendIdentitySaga),
        call(rootSendDocumentsSaga),
        call(rootSendEmailSaga),
        call(rootRecentTradesSaga),
        call(rootOrderBookSaga),
        call(rootHandleErrorSaga),
        call(rootCurrencyHistorySaga),
        call(rootUserActivitySaga),
    ]);
}
