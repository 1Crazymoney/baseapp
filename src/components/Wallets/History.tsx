import { History } from '@openware/components';
import * as moment from 'moment';
import * as React from 'react';
import { connect, MapDispatchToPropsFunction } from 'react-redux';
import { localeDate } from '../../helpers';
import {
    CurrencyHistory,
    fetchCurrencyHistory,
    resetCurrencyHistory,
    RootState,
    selectCurrencyHistory,
    selectCurrencyHistoryLoading,
    selectCurrentPage,
    selectFirstElemIndex,
    selectFullHistory,
    selectLastElemIndex,
    selectNextPageExists,
    selectPageCount,
} from '../../modules';
import { FailIcon } from './FailIcon';
import { NextPageIcon } from './NextPageIcon';
import { PreviousIcon } from './PreviousIcon';
import { SucceedIcon } from './SucceedIcon';


export interface HistoryProps {
    label: string;
    type: string;
    currency: string;
}

export interface ReduxProps {
    list: CurrencyHistory[];
    fetching: boolean;
    fullHistory: number;
    page: number;
    pageCount: number;
    firstElemIndex: number;
    lastElemIndex: number;
    nextPageExists: boolean;
}

interface DispatchProps {
    fetchCurrencyHistory: typeof fetchCurrencyHistory;
    resetCurrencyHistory: typeof resetCurrencyHistory;
}

export type Props = HistoryProps & ReduxProps & DispatchProps;

export class WalletTable extends React.Component<Props> {
    public componentDidMount() {
        const { type, currency } = this.props;
        this.props.fetchCurrencyHistory({ page: 0, currency, type, fullHistory: 0 });
    }

    public componentWillReceiveProps(nextProps) {
        const { type, currency } = this.props;
        if (nextProps.currency !== currency || nextProps.type !== type) {
            this.props.resetCurrencyHistory();
            this.props.fetchCurrencyHistory({ page: 0, currency: nextProps.currency, type, fullHistory: 0 });
        }
    }

    public componentWillUnmount() {
        this.props.resetCurrencyHistory();
    }

    public render() {
        const { label, list, fullHistory, firstElemIndex, lastElemIndex } = this.props;

        if (!list.length) {
            return null;
        }
        return (
            <div className="pg-history-elem">
                <div className="pg-history-elem__label">{label}</div>
                <History headers={['Date', 'Status', 'Amount']} data={this.retrieveData(list)}/>
                <div className="pg-history-elem__pagination">
                    <p>{firstElemIndex} - {lastElemIndex} of {fullHistory}</p>
                    <span className="pg-history__pagination-prev" onClick={this.onClickPrevPage}>
                        <PreviousIcon/>
                    </span>
                    <span className="pg-history__pagination-next" onClick={this.onClickNextPage}>
                        <NextPageIcon/>
                    </span>
                </div>
            </div>
        );
    }

    private onClickPrevPage = () => {
        const { page, type, currency, fullHistory } = this.props;
        if (page === 0) {
            return;
        }
        this.props.fetchCurrencyHistory({ page: page - 1, currency, type, fullHistory });
    };

    private onClickNextPage = () => {
        const { nextPageExists, page, type, currency, fullHistory } = this.props;
        if (!nextPageExists) {
            return;
        }
        this.props.fetchCurrencyHistory({ page: page + 1, currency, type, fullHistory });
    };

    private retrieveData = (list: CurrencyHistory[]) => {
        if (list.length === 0) {
            return [['There is no data to show...', '', '']];
        }
        return [...list]
            .sort((a: CurrencyHistory, b: CurrencyHistory) => {
                return moment(localeDate(a.created_at), 'DD/MM HH:mm') > moment(localeDate(b.created_at), 'DD/MM HH:mm') ? -1 : 1;
            })
            .map(item => {
                return [
                    moment(item.created_at).format('DD MMM YYYY'),
                    this.formatTxState(item.state),
                    item.amount,
                ];
            });
    };

    private formatTxState = (tx: string) => {
        const statusMapping = {
            succeed: <SucceedIcon/>,
            failed: <FailIcon/>,
            accepted: <SucceedIcon/>,
            collected: <SucceedIcon/>,
            canceled: <FailIcon/>,
            rejected: <FailIcon/>,
            processing: 'Pending',
            prepared: 'Pending',
            submitted: 'Pending',
        };
        return statusMapping[tx];
    };
}


export const mapStateToProps = (state: RootState): ReduxProps => ({
    list: selectCurrencyHistory(state),
    fetching: selectCurrencyHistoryLoading(state),
    fullHistory: selectFullHistory(state),
    page: selectCurrentPage(state),
    pageCount: selectPageCount(state),
    firstElemIndex: selectFirstElemIndex(state),
    lastElemIndex: selectLastElemIndex(state),
    nextPageExists: selectNextPageExists(state),
});

export const mapDispatchToProps: MapDispatchToPropsFunction<DispatchProps, {}> =
    dispatch => ({
        fetchCurrencyHistory: params => dispatch(fetchCurrencyHistory(params)),
        resetCurrencyHistory: () => dispatch(resetCurrencyHistory()),
    });

export const WalletHistory = connect(mapStateToProps, mapDispatchToProps)(WalletTable);
