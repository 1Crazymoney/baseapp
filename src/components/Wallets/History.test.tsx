import { shallow } from 'enzyme';
import * as React from 'react';
import { createStore } from 'redux';
import { rootReducer } from '../../modules';
import { mapDispatchToProps, mapStateToProps, WalletTable } from './History';


describe('WalletHistory', () => {
    let wrapper;
    const [spyReset, spyFetch] = [jest.fn(), jest.fn()];
    const reduxState = createStore(rootReducer).getState();

    const fakeHistory = [
        {
            id:566,
            currency:'btc',
            amount:'0.99',
            fee:'0.0',
            txid:'4516e174e7f04fafd14026c22d2bc288695aaa96f4b44518aa86ac7e27fc2458',
            created_at:'2018-12-03T17:13:58+01:00',
            confirmations:1,
            completed_at:'2018-12-03T17:14:56+01:00',
            state:'accepted',
        },
        {
            id:393,
            currency:'btc',
            amount:'0.001',
            fee:'0.0',
            txid:'dd5024e99c92aaa8787ed8273c8a6b635388eb4624d9cc1f8e04313dce843180',
            created_at:'2018-11-16T09:56:38+01:00',
            confirmations:0,
            completed_at:'2018-11-16T09:56:56+01:00',
            state:'canceled',
        },
    ];

    const props = {
        label: 'Deposits history',
        type: 'deposits',
        currency: 'btc',
        list: fakeHistory,
        fetching: false,
        fullHistory: 10,
        page: 0,
        pageCount: 2,
        firstElemIndex: 1,
        lastElemIndex: 6,
        nextPageExists: true,
        fetchCurrencyHistory: spyFetch,
        resetCurrencyHistory: spyReset,
    };

    beforeEach(() => {
        wrapper = shallow(<WalletTable {...props}/>);
    });

    it('fetchCurrencyHistory called', () => {
        expect(spyFetch).toHaveBeenCalled();
        //each test increases call +1, so it should be first (test componentDidMount call)
        expect(spyFetch).toHaveBeenCalledTimes(1);
        const args = { page: 0, currency: 'btc', type: 'deposits', fullHistory: 0 };
        expect(spyFetch).toHaveBeenCalledWith(args);
    });

    it('fetchCurrencyHistory called when tab changed', () => {
        wrapper.setProps({ currency: 'eth' });
        expect(spyFetch).toHaveBeenCalled();
        // each test increases call +1, so it should be first (test componentDidMount call)
        // 1 - prev test
        // 2 - componentDidMount
        // 3 - componentWillReceiveProps
        expect(spyFetch).toHaveBeenCalledTimes(3);
        const args = { page: 0, currency: 'eth', type: 'deposits', fullHistory: 0 };
        expect(spyFetch).toHaveBeenCalledWith(args);
        expect(spyReset).toHaveBeenCalledTimes(1);
    });

    it('should call resetCurrencyHistory during component unmount', () => {
        wrapper.unmount();
        expect(spyReset).toHaveBeenCalledTimes(2);
    });


    it('renders without crashing', () => {
        expect(wrapper).toBeDefined();
    });

    it('should have correct className', () => {
        expect(wrapper.hasClass('pg-history-elem'));
    });

    it('has label with right text', () => {
        expect(wrapper.find('.pg-history-elem__label').text()).toEqual('Deposits history');
    });

    it('has pagination info with right text', () => {
        expect(wrapper.find('p').text()).toEqual('1 - 6 of 10');
    });

    it('should test click on prev page', () => {
        const spyFetchPrevPage = jest.fn();
        const wrap = shallow(<WalletTable {...{...props, ...{ fetchCurrencyHistory: spyFetchPrevPage }}}/>);
        const prevButton = wrap.find('.pg-history__pagination-prev');
        // click when page equals 1
        prevButton.simulate('click');
        expect(spyFetchPrevPage).toHaveBeenCalledTimes(1);

        wrap.setProps({ page: 1 });
        // click when page equals 2
        prevButton.simulate('click');
        expect(spyFetchPrevPage).toHaveBeenCalledTimes(2);
    });

    it('should test click on next page', () => {
        const spyFetchNextPage = jest.fn();
        const wrap = shallow(<WalletTable {...{...props, ...{ fetchCurrencyHistory: spyFetchNextPage }}}/>);
        const prevButton = wrap.find('.pg-history__pagination-next');
        wrap.setProps({ nextPageExists: false });
        // click when page equals 1
        prevButton.simulate('click');
        expect(spyFetchNextPage).toHaveBeenCalledTimes(1);

        wrap.setProps({ nextPageExists: true });
        // click when page equals 2
        prevButton.simulate('click');
        expect(spyFetchNextPage).toHaveBeenCalledTimes(2);
        const args = { page: 0, currency: 'btc', type: 'deposits', fullHistory: 0 };
        expect(spyFetchNextPage).toHaveBeenCalledWith(args);
    });

    it('should matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('has redux context', () => {
        const dispatch = jest.fn();

        expect(mapStateToProps(reduxState).list).toEqual([]);
        expect(mapStateToProps(reduxState).fetching).toEqual(false);
        expect(mapStateToProps(reduxState).fullHistory).toEqual(0);
        expect(mapStateToProps(reduxState).page).toEqual(0);
        expect(mapStateToProps(reduxState).pageCount).toEqual(0);
        expect(mapStateToProps(reduxState).firstElemIndex).toEqual(1);
        expect(mapStateToProps(reduxState).lastElemIndex).toEqual(0);
        expect(mapStateToProps(reduxState).nextPageExists).toEqual(false);
        expect(mapDispatchToProps(dispatch, {}).fetchCurrencyHistory).toBeDefined();
        expect(mapDispatchToProps(dispatch, {}).resetCurrencyHistory).toBeDefined();
    });
});
