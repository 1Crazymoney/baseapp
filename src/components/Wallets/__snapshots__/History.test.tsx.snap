// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletHistory should matches snapshot 1`] = `
<div
  className="pg-history-elem"
>
  <div
    className="pg-history-elem__label"
  >
    Deposits history
  </div>
  <e
    data={
      Array [
        Array [
          "03 Dec 2018",
          <Unknown />,
          "0.99",
        ],
        Array [
          "16 Nov 2018",
          <Unknown />,
          "0.001",
        ],
      ]
    }
    headers={
      Array [
        "Date",
        "Status",
        "Amount",
      ]
    }
  />
  <div
    className="pg-history-elem__pagination"
  >
    <p>
      1
       - 
      6
       of 
      10
    </p>
    <span
      className="pg-history__pagination-prev"
      onClick={[Function]}
    >
      <Component />
    </span>
    <span
      className="pg-history__pagination-next"
      onClick={[Function]}
    >
      <Component />
    </span>
  </div>
</div>
`;
